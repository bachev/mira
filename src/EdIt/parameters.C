/*
 * Written by Thomas Pfisterer
 *
 * Copyright (C) 1997-2000 by the German Cancer Research Center (Deutsches
 *   Krebsforschungszentrum, DKFZ Heidelberg) and Thomas Pfisterer
 *
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 *
 */

#include "EdIt/parameters.H"


EDITParameters::EDITParameters()
{

  tag_parameters.tag_insert    = true;
  tag_parameters.tag_delete    = true;
  tag_parameters.tag_alter     = true;
  tag_parameters.tag_asterisk  = true;
  tag_parameters.tag_consensus = true;


  file_parameters.caf_input   = conditionalStrCpy("EdIt.in.caf");
  file_parameters.caf_output  = conditionalStrCpy("EdIt.out.caf");
  file_parameters.log_output  = conditionalStrCpy("EdIt.log");
  file_parameters.text_output = nullptr;
  file_parameters.html_output = nullptr;
  file_parameters.caf_void = false;

  strand_parameters.make_double_stranded     = true;
  strand_parameters.doublestrand_machinetype = false;
  strand_parameters.doublestrand_minscore = 85;
  strand_parameters.doublestrand_umgebung = 2;
  strand_parameters.mincoverage_single = 4;

  eval_parameters.min_solutions_checked  = 3;
  eval_parameters.max_score_over_best    = 20;
  eval_parameters.confirmation_threshold = 0.6;

  // eval_parameters.strict_evaluation      = true;
  eval_parameters.strict_evaluation      = false;
  eval_parameters.edit_quality_below     = -1;

  setEditByColumn(false);
  setEditByRegion(true);
  setEditBySmallRegions(false);

  command_edit   = true;   // Perform edit operations
  command_eval   = true;   // Evaluate hypotheses
  command_hypo   = true;   // Generate hypotheses

  eval_range.contig_id_von = -1;     // for all Contigs
  eval_range.contig_id_bis = -1;     // for all Contigs
  eval_range.contig_pos_von = -1;    // for all bases
  eval_range.contig_pos_bis = -1;    // for all bases

  verbose       = 0;
  show_progress = true;
  debug         = 0;

  help = false;   // print usage() instead of doing something

  log_cleared=false;
}


EDITParameters::~EDITParameters()
{
  delete [] file_parameters.caf_input;
  delete [] file_parameters.caf_output;
  delete [] file_parameters.log_output;
  delete [] file_parameters.html_output;
  delete [] file_parameters.text_output;
}


// return 0 if ok
int32 EDITParameters::setTagParams(ep_tag_parameters & tp)
{
  tag_parameters = tp;
  return 0;
}


ep_tag_parameters const & EDITParameters::getTagParams() const
{
  return tag_parameters;
}



void EDITParameters::logOperation(char *s) const
{
  static int32 count = 0;
  ofstream x;

  if (file_parameters.caf_void == false) {
    if(!log_cleared){
      x.open(file_parameters.log_output);
      x.close();
      bool & lc=const_cast<bool &>(log_cleared);
      lc=true;
    }
    x.open(file_parameters.log_output, ios::app);
    x << ++count << "\t" << s << endl;
    x.close();
  }
}




void EDITParameters::cmdLine2Stream(int argc, char **argv)
{
  stringstream tss;

  for(int32 i=1; i<argc; i++){
    //cout << "Doing " << i << "\t" << argv[i] << endl;
    tss << argv[i] << "-=BEGIN0=-";
  }

  parse(tss);
}


int32 EDITParameters::gimmeAnInt(FlexLexer * lexer)
{
  if(lexer->yylex() != EP_INT){
    cout << "Expected an int";
  }
  return atoi(lexer->YYText());
}


int32 EDITParameters::getConAnalyseMode(FlexLexer * lexer)
{
  int32 tmp=lexer->yylex();
  if(tmp==666){
    cout << "Non recognised am option";
    tmp=0;
  }
  return tmp;
}


bool EDITParameters::doContig(int32 id)
{
  // evaluate all contigs
  if (eval_range.contig_id_von < 0) return true;

  // evaluate contigs starting with...
  if (id >= eval_range.contig_id_von) {
    if (eval_range.contig_id_bis < 0) {
      return true;
    } else {
      return (id <= eval_range.contig_id_bis);
    }
  } else {
    return false;
  }
}


void EDITParameters::setFilenames(char *input, char *output, char *logfile)
{
  if (input != nullptr) {
    delete [] file_parameters.caf_input;
    file_parameters.caf_input = new char[strlen(input) + 1];
    strcpy(file_parameters.caf_input, input);
  }

  if (output != nullptr) {
    delete [] file_parameters.caf_output;
    file_parameters.caf_output = new char[strlen(output) + 1];
    strcpy(file_parameters.caf_output, output);
  }

  if (logfile != nullptr) {
    delete [] file_parameters.log_output;
    file_parameters.log_output = new char[strlen(logfile) + 1];
    strcpy(file_parameters.log_output, logfile);
  }

  file_parameters.caf_void = (logfile == nullptr);
}



void EDITParameters::parse(istream & is)
{
  FlexLexer* lexer = new EPFlexLexer(&is);

  int yyretcode=-1;

  while(yyretcode!=0){
    yyretcode=lexer->yylex();

    switch(yyretcode){
    case 0: break;                              // do nothing, eof

    case EP_ALTER_TAG :{
      tag_parameters.tag_alter = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: alter tag set to "
	   << tag_parameters.tag_alter << endl;
      break;
    }

    case EP_DELETE_TAG: {
      tag_parameters.tag_delete = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: delete tag set to "
	   << tag_parameters.tag_alter << endl;
      break;
    }

    case EP_INSERT_TAG: {
      tag_parameters.tag_insert = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: insert tag set to "
	   << tag_parameters.tag_alter << endl;
      break;
    }

    case EP_CONSENSUS_TAG: {
      tag_parameters.tag_consensus = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: consensus tag set to "
	   << tag_parameters.tag_consensus << endl;
      break;
    }

    case EP_VERBOSE: {
      verbose = gimmeAnInt(lexer);
      cout << "PARAMETER: verbose mode set to " << verbose << endl;
#ifdef RUNONLY
      cout << "           compiled as run-only; some output ignored" << endl;
#endif
      break;
    }

    case EP_DEBUG: {
      debug = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: debug mode set to " << debug << endl;
      break;
    }

    case EP_ALL_TAGS: {
      tag_parameters.tag_insert    = true;
      tag_parameters.tag_delete    = true;
      tag_parameters.tag_alter     = true;
      tag_parameters.tag_consensus = true;
      cout << "PARAMETER: insert, delete, alter, consensus tag set to 1."
	   << endl;
      break;
    }

   case EP_NO_TAGS: {
      tag_parameters.tag_insert    = false;
      tag_parameters.tag_delete    = false;
      tag_parameters.tag_alter     = false;
      tag_parameters.tag_consensus = false;
      cout << "PARAMETER: insert, delete, alter, consensus tag set to 0."
	   << endl;
      break;
    }

    case EP_DELETE_ASTERISK: {
      tag_parameters.tag_asterisk = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: delete asterisk columns set to "
	   << tag_parameters.tag_asterisk << endl;
      break;
    }

    case EP_caf_in_file: {
      delete [] file_parameters.caf_input;
      file_parameters.caf_input = new char[lexer->YYLeng() + 1];
      strcpy(file_parameters.caf_input, lexer->YYText());
      cout << "PARAMETER: input file : "
	   << file_parameters.caf_input << endl;
      break;
    }

    case EP_caf_out_file: {
      int32 l;

      delete [] file_parameters.caf_output;
      file_parameters.caf_output = new char[lexer->YYLeng() + 1];
      strcpy(file_parameters.caf_output, lexer->YYText());
      cout << "PARAMETER: output file : "
	   << file_parameters.caf_output << endl;
      break;
    }

   case EP_log_out_file: {
      int32 l;

      delete [] file_parameters.log_output;
      file_parameters.log_output = new char[lexer->YYLeng() + 1];
      strcpy(file_parameters.log_output, lexer->YYText());
      cout << "PARAMETER: log file : "
	   << file_parameters.log_output << endl;
      break;
    }

    case EP_html_out_file: {
      int32 l;

      delete [] file_parameters.html_output;
      file_parameters.html_output = new char[lexer->YYLeng() + 1];
      strcpy(file_parameters.html_output, lexer->YYText());
      cout << "PARAMETER: html file : "
	   << file_parameters.html_output << endl;
      break;
    }


    case EP_text_out_file: {
      int32 l;

      delete [] file_parameters.text_output;
      file_parameters.text_output = new char[lexer->YYLeng() + 1];
      strcpy(file_parameters.text_output, lexer->YYText());
      cout << "PARAMETER: text file : "
	   << file_parameters.text_output << endl;
      break;
    }


    case EP_outfile_void: {
      file_parameters.caf_void = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: create caf output: "
	   << file_parameters.caf_void << endl;
      break;
    }


    case EP_make_double: {
      strand_parameters.make_double_stranded = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: make double-stranded "
	   << strand_parameters.make_double_stranded << endl;
      break;
    }

    case EP_use_machinetype: {
      strand_parameters.doublestrand_machinetype = ((gimmeAnInt(lexer) == 1));
      cout << "PARAMETER: make double-stranded if different sequencers "
	   << strand_parameters.doublestrand_machinetype << endl;
      break;
    }

    case EP_doublestrand_minscore: {
      int32 score =  gimmeAnInt(lexer);
      if (score <= 100 && score >= 0) {
	strand_parameters.doublestrand_minscore = score;
      } else {
	cerr << "ERROR: minscore doublestranding outside [0,100]!\n";
      }
      break;
    }

    case EP_mincoverage_single: {
      int32 coverage = gimmeAnInt(lexer);
      if (coverage <= 100 && coverage >= -1) {
	strand_parameters.mincoverage_single = coverage;
	cout << "PARAMETER: minimum coverage for single stranded regions: "
	     << coverage << endl;
      } else {
	cerr << "ERROR: parameter 'cover' outside [-1,100]!\n";
      }
      break;
    }

    case EP_ANID: {
      cout <<  "PARAMETER: ERROR - unknown option: "
	   << lexer->YYText() <<endl;
      cout.flush();
      break;
    }

    case EP_DO_NOP: {
      cout << "PARAMETER: do nothing (read the caf-file)" << endl;
      command_hypo = command_eval = command_edit = false;
      break;
    }

    case EP_DO_ALL: {
      cout << "PARAMETER: do all (default)" << endl;
      command_hypo = command_eval = command_edit = true;
      break;
    }

    case EP_DO_HYPO: {
      cout << "PARAMETER: generate hypotheses only (no eval, no edit)"
	   << endl;
      command_hypo = true;
      command_eval = command_edit = false;
      break;
    }

    case EP_DO_EVAL: {
      cout << "PARAMETER: evaluate hypotheses but do not edit" << endl;
      command_hypo = command_eval = true;
      command_edit = false;
      break;
    }

    case EP_DO_CONTIG: {
      eval_range.contig_id_von = gimmeAnInt(lexer);
      eval_range.contig_id_bis = eval_range.contig_id_von ;

      cout << "PARAMETER: edit single contig - "
	   << eval_range.contig_id_von << endl;

      setDoAll();

      break;
    }

    case EP_DO_RANGE: {
      eval_range.contig_pos_von = gimmeAnInt(lexer);
      eval_range.contig_pos_bis = gimmeAnInt(lexer) ;

      cout << "PARAMETER: eval range von "
	   << eval_range.contig_pos_von << " - "
	   << eval_range.contig_pos_bis << endl;

      break;
    }

    case EP_DO_STRICT: {
      cout << "PARAMETER: strict evaluation ON" << endl;
      eval_parameters.strict_evaluation = true;
      break;
    }

    case EP_DO_NOSTRICT: {
      cout << "PARAMETER: strict evaluation OFF" << endl;
      eval_parameters.strict_evaluation = false;
      break;
    }

    case EP_DO_THRESHOLD: {
      setConfirmationThreshold((float)(gimmeAnInt(lexer))/100.0) ;
      cout << "PARAMETER: confirmation threshold "
	   << getConfirmationThreshold() << endl;
      break;
    }

    case EP_DO_LOWQUAL: {
      setEditLowQualityBases(gimmeAnInt(lexer));
      cout << "PARAMETER: edit if base quality below "
	   << getLowQualityThreshold() << endl;
      break;
    }

    case EP_DO_REGIONS: {
      setEditByRegion(gimmeAnInt(lexer) != 0);
      cout << "PARAMETER: edit by fault region is "
	   << getEditByRegion() << endl;
      break;
    }

    case EP_DO_COLUMNS: {
      setEditByColumn(gimmeAnInt(lexer) != 0);
      cout << "PARAMETER: edit by column is "
	   << getEditByColumn() << endl;
      break;
    }

    case EP_DO_SMALL_REGIONS: {
      setEditBySmallRegions(gimmeAnInt(lexer) != 0);
      cout << "PARAMETER: edit by small regions is "
	   << getEditBySmallRegions() << endl;
      break;
    }

    case EP_HELP: {
      setHelp(true);
      break;
    }

    default:{
	cout <<  "PARAMETER: ERROR - recognised but not caught: "
	     << lexer->YYText() <<endl;
	cout.flush();
    }
    }

  }

  delete lexer;

}



ostream &operator<<(ostream &ostr, EDITParameters const &i)
{
  ostr << "\nEdIt parameters : " << endl;
  ostr << "  Edit by:";

  if (i.getEditByRegion()) { ostr << " 'regions'"; }
  if (i.getEditBySmallRegions()) { ostr << " 'small regions'"; }
  if (i.getEditByColumn()) { ostr << "'columns'"; }
  ostr << "." << endl;

  ostr << "  Edit single contig: " << i.eval_range.contig_id_von << endl;
  ostr << "  Check at least " << i.eval_parameters.min_solutions_checked
       << " solutions until " << i.eval_parameters.max_score_over_best
       << " % over best score." << endl;

  ostr << "  Threshold for network evaluation: "
       << i.eval_parameters.confirmation_threshold
       << "." << endl;
  ostr << "  Edit quality below "
       << i.eval_parameters.edit_quality_below << endl;
  ostr << "  Strict evaluation :" ;
  if (i.eval_parameters.strict_evaluation) {
    ostr << " YES."  << endl;
  } else {
    ostr << " NO." << endl;
  }

  ostr << "  Tags : ";
  if (i.tag_parameters.tag_insert)    { ostr << " 'insert'"; }
  if (i.tag_parameters.tag_delete)    { ostr << " 'delete'"; }
  if (i.tag_parameters.tag_alter)     { ostr << " 'alter'"; }
  if (i.tag_parameters.tag_consensus) { ostr << " 'consensus'"; }
  if (i.tag_parameters.tag_asterisk)  { ostr << " 'asterisk'"; }
  ostr << "." << endl;

  ostr << "  Files: ";
  ostr << " " << i.file_parameters.caf_input << " ==> "
       << i.file_parameters.caf_output
       << "     log: " << i.file_parameters.log_output << endl;

  ostr << endl;
  return ostr;
}
