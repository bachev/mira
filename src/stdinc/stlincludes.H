/*
 * Written by Bastien Chevreux (BaCh)
 *
 * Copyright (C) 2003 and later by Bastien Chevreux
 *
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 *
 */

#ifndef STL_H
#define STL_H

// special function to really free memory of a STL vector
// not really pretty, but until STL defines an own way ...
template <class T>
void nukeSTLContainer(T&v)
{
  T tmp;
  v.swap(tmp);
}

//#include <algorithm>
//#include <bitset>
//#include <complex>
//#include <deque>
//#include <exception>
//#include <fstream>
//#include <functional>
//#include <ios>
//#include <iosfwd>
//#include <iostream>
//#include <istream>
//#include <iterator>
//#include <limits>
//#include <list>
//#include <locale>
//#include <map>
//#include <memory>
//#include <new>
//#include <numeric>
//#include <ostream>
//#include <queue>
//#include <set>
//#include <sstream>
//#include <stack>
//#include <stdexcept>
//#include <streambuf>
//#include <string>
//#include <typeinfo>
//#include <utility>
//#include <valarray>
//#include <vector>

//#if __GNUC__ >= 3
//#include <ext/hash_map>
//using namespace __gnu_cxx;          // hashes are in this namespace of gcc3
//#endif

//#include <ext/slist>

#endif
