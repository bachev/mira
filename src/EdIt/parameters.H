/*
 * Written by Thomas Pfisterer
 *
 * Copyright (C) 1997-2000 by the German Cancer Research Center (Deutsches
 *   Krebsforschungszentrum, DKFZ Heidelberg) and Thomas Pfisterer
 *
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
 *
 */


#ifndef _edit_parameters_H
#define _edit_parameters_H

#include "stdinc/defines.H"
#include "stdinc/stlincludes.H"

#include <stdlib.h>

#include "errorhandling/errorhandling.H"
#include "util/misc.H"

#include <sstream>

#include <fstream>
//#include <algo.h>                     // STL min max template
#include "EdIt/parameters_tokens.h"

#undef yyFlexLexer
#define yyFlexLexer EPFlexLexer
#include <FlexLexer.h>



// 	$Id$



struct ep_tag_parameters
{
  bool tag_alter;
  bool tag_insert;
  bool tag_delete;
  bool tag_asterisk;
  bool tag_consensus;
};


struct ep_file_parameters
{
  char *caf_input;
  char *caf_output;
  char *log_output;
  char *text_output;
  char *html_output;
  bool caf_void;
};



struct ep_strand_parameters
{
  bool  make_double_stranded;
  bool  doublestrand_machinetype;   // verschiedene sequenziermaschinen = doublestranded
  int32 doublestrand_minscore;
  int32 doublestrand_umgebung;
  int32 mincoverage_single;
};


struct ep_evaluation_parameters
{
  int32  min_solutions_checked;    // minimum number of solutions tested
  // threshold: evaluate solutions if their score is less than x(%) over
  // higher than the score of the best solution
  int32  max_score_over_best;
  float  confirmation_threshold;   // KNN result threshold to confirm hypo.
  bool   strict_evaluation;        // Edit only when completely confirmed.
  // Edit all bases below this quality value even without trace confirmation
  // if set to -1 (default) no automatic editing of low quality bases is done
  int32  edit_quality_below;
  bool   edit_by_columns;
  bool   edit_by_regions;
  bool   edit_by_small_regions;
};



struct ep_evaluation_range
{
  int32 contig_id_von;          // edit contigs starting from
  int32 contig_id_bis;          // until...
  int32 contig_len_von;         // edit contigs longer than...
  int32 contig_len_bis;         //              shorter than...
  int32 contig_pos_von;         // edit contig between
  int32 contig_pos_bis;         //              and...
};



class EDITParameters
{
private:

  ep_tag_parameters          tag_parameters;
  ep_file_parameters         file_parameters;
  ep_strand_parameters       strand_parameters;
  ep_evaluation_parameters   eval_parameters;
  ep_evaluation_range        eval_range;

  bool  command_hypo;
  bool  command_eval;
  bool  command_edit;

  int32 verbose;
  bool  show_progress;
  bool  debug;

  bool  log_cleared;

  bool  help;    // parameter "help" was set

  //Functions
private:
  void  foolCompiler();
  int32 gimmeAnInt(FlexLexer * lexer);
  int32 getConAnalyseMode(FlexLexer * lexer);

public:
  EDITParameters();
  ~EDITParameters();

  ep_tag_parameters  const & getTagParams() const ;

  void cmdLine2Stream(int argc, char **argv);
  void parse(istream & is);

  int32 setTagParams(ep_tag_parameters & ap);
  void  logOperation(char *s) const;

  void  setFilenames(char *input, char *output, char *logfile);
  float getConfirmationThreshold() const
    { return eval_parameters.confirmation_threshold; }
  void  setConfirmationThreshold(const float f)
    { eval_parameters.confirmation_threshold = f; }
  bool getStrictEvaluation()
    { return eval_parameters.strict_evaluation; }
  void  setStrictEvaluation(const bool mode)
    { eval_parameters.strict_evaluation = mode; }
  bool  getEditLowQualityBases()
    { return (eval_parameters.edit_quality_below > 0); }
  int32 getLowQualityThreshold()
    { return eval_parameters.edit_quality_below; }
  void  setEditLowQualityBases(int32 quality_threshold)
    { eval_parameters.edit_quality_below = quality_threshold; }
  bool  getEditByRegion() const
    { return eval_parameters.edit_by_regions; }
  void  setEditByRegion(bool b)
    { eval_parameters.edit_by_regions = b; }
  bool  getEditByColumn() const
    { return eval_parameters.edit_by_columns; }
  void  setEditByColumn(bool b)
    { eval_parameters.edit_by_columns = b; }
  bool  getEditBySmallRegions() const
    { return eval_parameters.edit_by_small_regions; }
  void  setEditBySmallRegions(bool b)
    { eval_parameters.edit_by_small_regions = b; }

  bool   tagInsertOperations() const
    { return tag_parameters.tag_insert; };
  bool   tagDeleteOperations() const
    { return tag_parameters.tag_delete; };
  bool   tagAlterOperations() const
    { return tag_parameters.tag_alter; };
  bool   tagDeleteAsteriskColumns() const
    { return tag_parameters.tag_asterisk; };
  bool   tagConsensus() const
    { return tag_parameters.tag_consensus; };

  char * getCafInputFileName()  const
    { return file_parameters.caf_input; };
  char * getCafOutputFileName() const
    { return file_parameters.caf_output; };
  bool   createCafOutput() const
    { return file_parameters.caf_void == false; }
  bool   createHtmlOutput() const
    { return file_parameters.html_output != nullptr; }
  char*  getHtmlOutputFileName() const
    { return file_parameters.html_output; }
  bool   createTextOutput() const
    { return file_parameters.text_output != nullptr; }
  char*  getTextOutputFileName() const
    { return file_parameters.text_output; }

  bool   makeDoubleStranded() const
    { return strand_parameters.make_double_stranded; };
  void   makeDoubleStranded(const bool ds)
    { strand_parameters.make_double_stranded = ds; };
  int32  getDoubleStrandMinScore() const
    { return strand_parameters.doublestrand_minscore; };
  int32  getDoublestrandUmgebung() const
    { return strand_parameters.doublestrand_umgebung; };
  int32  getMinCoverage() const
    { return strand_parameters.mincoverage_single; };

  int32  getMinSolutionsChecked() const
    { return eval_parameters.min_solutions_checked; };
  int32  getMaxScoreOverBest() const
    { return eval_parameters.max_score_over_best; };

  bool doSomething() const
    { return command_hypo == true; }
  bool doHypo() const
    { return command_hypo == true; }
  bool doEval() const
    { return command_eval == true; }
  bool doEdit() const
    { return command_edit == true; }


  bool  doContig(int32 id);
  int32 evalAt(int32 pos)
    {
      if (eval_range.contig_pos_von == -1) return true;
      return (pos >= eval_range.contig_pos_von &&
	      pos <= eval_range.contig_pos_bis);
    }
  void  setEvalRange(int32 pos_von, int32 pos_bis)
    {
      eval_range.contig_pos_von = pos_von;
      eval_range.contig_pos_bis = pos_bis;
    }

  bool isVerbose() const
    { return verbose > 0; }

  bool isVerbose(int32 i) const
    { return verbose > i; }

  bool isDebug() const
    { return debug == true; }

  void setVerbose(bool isitVerbose)
    { if (isitVerbose) { verbose = 1; } else { verbose = 0; }}

  void setVerbose(int32 verbmode)
    { verbose = verbmode; }

  void setShowProgress(bool b)
    { show_progress = b; }

  bool getShowProgress()
    { return show_progress; }



  void setDoNothing() {
    command_hypo = false;
    command_eval = false;
    command_edit = false;
  }

  void setDoHypo() {
    command_hypo = true;
    command_eval = true;
    command_edit = false;
  }

  void setDoEval() {
    command_hypo = true;
    command_eval = true;
    command_edit = false;
  }

  void setDoAll() {
    command_hypo = true;
    command_eval = true;
    command_edit = true;
  }

  bool getHelp() { return help; }
  void setHelp(bool h) { help = h; }

  friend ostream &operator<<(ostream &ostr, EDITParameters const &i);
};

#endif
