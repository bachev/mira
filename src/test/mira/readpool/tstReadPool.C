#include <limits.h>
#include <iostream>

#include "stdinc/stlincludes.H"
#include "mira/parameters.H"
#include "mira/readpool.H"


#define NAMELEN 40


char error[][NAMELEN]= { 
  {"fofnerror.txt"}
};
char normal[][NAMELEN]= { 
  {"fofn.txt"}
};
char straindata[][NAMELEN]= { 
  {"fofn.txt"}
};

int main(int argc, char ** argv)
{
  FUNCSTART("main");

  try{
    vector<MIRAParameters> Pv;
    MIRAParameters::setupStdMIRAParameters(Pv);
    MIRAParameters::parse(argc, argv, Pv);


   {
     ReadPool rp(&Pv);
     
     cout << "Testing error in fofn  operations\n";
     for(uint32 i=0; i < sizeof(error)/NAMELEN; i++ ) {
       cout << "############ 1-" << i << endl;
       try{
	 cout << "Testing " << error[i] << "\n------------------------------------------------" << endl;
	 uint32 dummy=0;
	 rp.loadEXPs(error[i],1,dummy);
	 cout << "Done.";
	 cout << "Readpool contains " << rp.size() << " reads." << endl;
       }
       catch(Notify n){
	 n.setGravity(Notify::WARNING);
	 n.handleError("main()");
       }
       cout << "\n\n";
     }
   }

    {
      ReadPool rp(&Pv);
      
      cout << "Testing normal EXP load operations\n";
      for(uint32 i=0; i < sizeof(normal)/NAMELEN; i++ ) {
	cout << "############ 2-" << i << endl;
	try{
	  cout << "Testing " << normal[i] << "\n------------------------------------------------" << endl;
	  uint32 dummy=0;
	  rp.loadEXPs(normal[i],1,dummy);
	  cout << "Done.";
	  cout << "Readpool contains " << rp.size() << " reads." << endl;
	}
	catch(Notify n){
	  n.setGravity(Notify::WARNING);
	  n.handleError("main()");
	}
	cout << "\n\n";
      }
    }

    {
      ReadPool rp(&Pv);
      
      cout << "Testing straindata load operations\n";
      for(uint32 i=0; i < sizeof(straindata)/NAMELEN; i++ ) {
	cout << "############ 3-" << i << endl;
	try{
	  cout << "Testing " << straindata[i] << "\n------------------------------------------------" << endl;
	  uint32 dummy=0;
	  rp.loadEXPs(straindata[i],1,dummy);
	  cout << "Done.";
	  cout << "Readpool contains " << rp.size() << " reads." << endl;
	}
	catch(Notify n){
	  n.setGravity(Notify::WARNING);
	  n.handleError("main()");
	}
	cout << "\n\n";
      }
    }
  }
  catch(Notify n){
    n.handleError("main");
  }
  cout << "\n\n";

  cout << "Tests ended.\n";

  FUNCEND();
  return 0;
}


