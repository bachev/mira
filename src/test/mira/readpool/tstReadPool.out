Testing error in fofn  operations
############ 1-0
Testing fofnerror.txt
------------------------------------------------
Loading file of filenames: fofnerror.txt
File U13a03f09.t1 is present more than once.
done.

Warning: "Some reads lead to unrecoverable error: duplicate names. Aborting, see stderr for further information."
->Thrown: void ReadPool::loadEXPs(const char * fofn)  
->Caught: main()


Testing normal EXP load operations
############ 2-0
Testing fofn.txt
------------------------------------------------
Loading file of filenames: fofn.txt
done.
Loading EXP files: 
 [0%] .......... [10%] .......... [20%] .......... [30%] .......... [40%] .......... [50%] .......... [60%] .......... [70%] .......... [80%] .......... [90%] .......... [100%] 
Done.
There haven been 5 files given, 5 of which have loaded ok.
Adjusting sequencing vector clips from tags:
 [0%] .......... [10%] .......... [20%] .......... [30%] .......... [40%] .......... [50%] .......... [60%] .......... [70%] .......... [80%] .......... [90%] .......... [100%] Done.Readpool contains 5 reads.


Testing straindata load operations
############ 3-0
Testing fofn.txt
------------------------------------------------
Loading file of filenames: fofn.txt
done.
Loading EXP files: 
 [0%] .......... [10%] .......... [20%] .......... [30%] .......... [40%] .......... [50%] .......... [60%] .......... [70%] .......... [80%] .......... [90%] .......... [100%] 
Done.
There haven been 5 files given, 5 of which have loaded ok.
Adjusting sequencing vector clips from tags:
 [0%] .......... [10%] .......... [20%] .......... [30%] .......... [40%] .......... [50%] .......... [60%] .......... [70%] .......... [80%] .......... [90%] .......... [100%] Building hash table ... done.
U13a03f07.t1 has id 1 in readpool, setting strain to bla
U13a01g08.t1 has id 0 in readpool, setting strain to bla
not_an_EXP from straindata not found in readpool.
U13a05h03.t1 has id 3 in readpool, setting strain to blubb
U13a03f09.t1 has id 2 in readpool, setting strain to blubb
U13a08c07.t1 has id 4 in readpool, setting strain to ble

Read 6 straindata lines.
Done.Readpool contains 5 reads.




Tests ended.
